### PR TITLE
Clarify vertical spacing

### LINKED ISSUE #1374: Strange behavior when `spacebeneathtext` is set to negative value
In the following MWE, the score, as a whole, gets pushed down as a result of the negative value for `spacebeneathtext` but the lines do not get any closer together.

```
\documentclass[12pt]{article}

\usepackage{gregoriotex}

\begin{document}
text needed to show how a negative spacebeneathtext value will move the whole score

\grechangedim{spacebeneathtext}{-2cm}{scalable}
\gabcsnippet{(c3)Herr,(e) ich(f) {traue auf}(hr0) || <ul>dich</ul>,(i) { }(fr) { }(f.) *(;) (z) {lass mich nimmermehr zu-}(hr0) || schan(g)den(h) | wer(f) {-}(fr) den.(ee) (::)}

\end{document}
```

What should the behavior here be?  Should the lines be pushed into overlapping?  Or perhaps we need to simply add a check to make sure that `spacebeneathtext` is not being set to a negative value?

### PR DESCRIPTION
**Pull request by davidweichiang**
This patch addresses #1631 by making two changes to vertical spacing and adding more detailed documentation about vertical spacing. The two changes unfortunately will break some existing scores, but hopefully make spacing less confusing and easier to extend in the future.

A line without above-lines text looks like this. In all these pictures, the following lengths have been changed from their defaults:

```
\grechangedim{spaceabovelines}{0.2cm}{scalable}
\grechangedim{spacebeneathtext}{0.2cm}{scalable}
\grechangedim{abovelinestextraise}{0.2cm}{scalable}
```

image

But with above-lines text, the picture becomes more complicated:

image

The total height is the maximum of the two heights shown.

The first problem is that the docs say that `abovelinestextraise` is the "Height of the text above the note line." To me, that sounds like the distance from the top staff line to the baseline of the above-lines text. But there's actually an extra `interstafflinespace` in there, which I'm guessing was an off-by-one error.

The second problem is that the docs say that `abovelinestextheight` "must be bigger than the text for it to be taken into consideration," but in this example, one would have to increase `abovelinestextheight` by quite a bit more than that before seeing any effect. The reason is that the total height computation leaves out `abovelinestextraise`, and I'm guessing this was just an oversight.

Under this patch, the picture becomes:

image

It makes two changes:

1. `abovelinestextraise` is measured from the top staff line. The default `abovelinestextraise` is changed from -0.1cm to 1.7351cm to produce the same result as before (all tests pass). But any code that sets `abovelinestextraise` explicitly would have to be corrected.

2. `abovelinestextraise` is included in the computation of the total height. Decreasing the default `abovelinestextheight` to compensate (0.12649cm) makes all tests pass, but I think the current default of 0.3 cm is good, and adds a little breathing room above the above-lines text.

The new documentation is on pages 77-82 of GregorioRef.pdf. It goes into more detail about additional space above/below the staff, and about `baselineskip` and `lineskip`.

Closes #1374, #1375, #1631.

### CODE REVIEW

**Comment by davidweichiang** [comment]
Closes #1660.

File: doc/Command_Index_User.tex [hunk 1]
```diff
@@ -1885,25 +1857,301 @@ \subsection{Distances}\label{distances}
...
+\end{gdimension}
+
+\begin{gdimension}{translationheight}
+The space between the baseline of the text and the baseline of the translation.
+\end{gdimension}
+
+\begin{gdimension}{spacebeneathtext}
+The space beneath the text (and translation, if any).
+\end{gdimension}
+
```
Comment (rpspringuel) [inline doc/Command_Index_User.tex:1981]: When I create the pdf, the entry for `spacebeneathtext` is forced to a new page when there's plenty of room for it on the previous page.  We need to do something to encourage page breaking at this point.

File: doc/Command_Index_User.tex [hunk 2]
```diff
@@ -1885,25 +1857,301 @@ \subsection{Distances}\label{distances}
...
+A legacy option is to change the amount of additional bottom space per step by setting \verb|\gresetnoteadditionalspacelinestext{manual}| and the following:
+
+\begin{gdimension}{noteadditionalspacelinestext}
+The per-step additional space between lines and the bottom of the text, when using \verb|\gresetnoteadditionalspacelinestext{manual}|.
+\end{gdimension}
+
+\paragraph{Vertical spacing between staves}
+
+The standard rules for paragraph and line spacing apply to scores and staves.
```
Comment (rpspringuel) [inline doc/Command_Index_User.tex]: While the standard rules apply, the below listed distances affect how the rules are applied inside a score.  I think that distinction should be spelled out for users who might not be familiar with how the TeX rules work.  Maybe something like:

> The standard rules for paragraph and line spacing apply to scores and staves.  However, the distances which control those rules are different inside a score from in the regular text.  The values listed below will be used inside a score and can be changed with `\grechangedim` without affecting how text outside of a score is typeset.

**Comment by rpspringuel** [comment]
These changes affect several tests.  I didn't see anything that stood out to me as problematic, but I'd like you to check that the changes are what you're expecting.

**Comment by davidweichiang** [comment]
These tests fail for me:

gabc-output/glyphs/clear.gabc
gabc-output/glyphs/nabc-edge-cases.gabc
gabc-output/glyphs/above_lines_text.gabc
gabc-output/FactusEst.gabc
gabc-output/bugs/fix-1613.gabc
gabc-output/bugs/fix-1015.gabc
gabc-output/bugs/fix-347.gabc
gabc-output/bugs/test98.gabc
gabc-output/soft-alterations.gabc
gabc-output/heights_legacy.gabc
gabc-output/heights.gabc
gabc-output/high_clef.gabc
tex-output/bugs/fix-1614/fix-1614.tex
tex-output/hideabovelinestext/hideabovelinestext.tex
tex-output/bar-spacing/end-of-line.tex
tex-output/bar-spacing-old/bar-spacing-old.tex
tex-output/bar-spacing/bar-spacing-new.tex

All of these have either above-lines text or NABC, and the differences are all slight increases in the space above-lines text or NABC. As mentioned above, there's a setting for abovelinestextheight that makes all the tests pass. But I think it's better to have that little extra space, as long as you're okay with changing existing scores.

File: doc/Command_Index_User.tex (hunk 2 above)
Comment (davidweichiang) [reply doc/Command_Index_User.tex]: Done

File: doc/Command_Index_User.tex (hunk 1 above)
Comment (davidweichiang) [reply doc/Command_Index_User.tex:1981]: That page break was really weird. I inserted a \goodbreak to fix it and hope that's ok.

Comment (rpspringuel) [reply doc/Command_Index_User.tex:1981]: Looks good now.

File: doc/Command_Index_User.tex [hunk 3]
```diff
@@ -1885,25 +1857,316 @@ \subsection{Distances}\label{distances}
...
+\verb|\grechangedim{abovelinestextheight}{0.1cm}{scalable}| \\
+\grechangedim{abovelinestextraise}{0.8cm}{scalable}%
+\verb|\grechangedim{abovelinestextraise}{0.8cm}{scalable}|%
+\gresetlyriccentering{syllable}%
+\begin{center}
+\rowcolors{1}{white}{white}%
+\begin{tabular}{ccccc}
+  \lines{\nabcsnippet{m|vi[0 steps](m|vi)}} &
+  \lines{\nabcsnippet{m_1|vi[0 steps](m_1|vi)}} &
+  \lines{\nabcsnippet{m'1|vi[1 step](m'1|vi)}} &
```
Comment (rpspringuel) [inline doc/Command_Index_User.tex]: The straight quote in the lyrics of this and the next snippet are being converted to curly quotes in the output.  Please make use of `\textsinglequote` in the lyrics to get the quote to print correctly.

**Comment by rpspringuel** [comment]
> All of these have either above-lines text or NABC, and the differences are all slight increases in the space above-lines text or NABC. As mentioned above, there's a setting for abovelinestextheight that makes all the tests pass. But I think it's better to have that little extra space, as long as you're okay with changing existing scores.

As long as the changes are expected given the changes you're submitting here, I don't have a problem with them.  I just wanted to make sure the changes in the tests were all in accord with the expected differences these changes would cause.

File: doc/Command_Index_User.tex (hunk 3 above)
Comment (davidweichiang) [reply doc/Command_Index_User.tex]: Done (and changed font to match similar examples elsewhere)

File: doc/Command_Index_User.tex [hunk 4]
```diff
@@ -2093,11 +2099,11 @@ \subsubsection{Vertical spacing}
 \end{gdimension}
 
 \begin{gdimension}{lineskiplimit}
-The effective \verb=\lineskiplimit= inside of a score. If two consecutive staves have baselines \texttt{baselineskip} apart and the distance between the first staff's bottomline and the second staff's topline is less than this, then \texttt{lineskip} is added between them.
+The effective \verb=\lineskiplimit= inside of a score. If two consecutive staves have baselines \texttt{baselineskip} apart and the distance between the first staff's bottome and the second staff's top is less than this, then \texttt{lineskip} is added between them.
```
Comment (rpspringuel) [inline doc/Command_Index_User.tex]: s/bottome/bottom/

Comment (davidweichiang) [reply doc/Command_Index_User.tex]: Done

(I realized that "topline" means something different than how I was using it, so I changed "topline" and "bottomline" to "top" and "bottom".)

**Comment by davidweichiang** [comment]
Yes, the changed tests are all expected given the changes that are made to the meanings of `abovelinestextraise` and `abovelinestextheight`.

**Comment by davidweichiang** [comment]
@rpspringuel @eschwab Is there anything else you'd like to see fixed in this PR? I will have a second PR that is waiting on this one; maybe I'll go ahead and open that one as a draft.

**Comment by davidweichiang** [comment]
There's a paragraph in the new section that says that if you want to ensure space above the top note, you should set `spaceabovelines` to something greater than 0.41cm (the default is 0cm). Should the default `spaceabovelines` be increased to 0.41cm?

cf. https://groups.google.com/g/gregorio-users/c/_K5qfmGVunI/m/OwOCOC5cDgAJ

**Comment by rpspringuel** [comment]
I'm satisfied with the documentation updates.

On the issue of defaults, I'm all for defaults that "make sense" or "seem natural", but understand that this is hard to define because different people have different opinions on what qualifies.  With all the changes that have been made recently to the spacing (including the PRs not yet merged) it may be time to reconsider our defaults, but I would make that a separate PR.

**Comment by davidweichiang** [comment]
> I'm satisfied with the documentation updates.

Great, and the changes to the meanings of `abovelinestextraise` and `abovelinestextheight` as well?

> I would make [changing defaults] a separate PR.

That makes sense to me.

**Comment by rpspringuel** [comment]
> Great, and the changes to the meanings of abovelinestextraise and abovelinestextheight as well?

I'm good with those redefined meanings.  You've managed to simplify things and explain them in a way that should make things much easier to understand.  The vertical spacings as they existed before this PR were always confusing to me (and plenty of others).
